posthog/hogql/ast.py

```diff
@@ -424,20 +445,47 @@ def is_nullable(self, context: HogQLContext) -> bool:
         return True
 
     def resolve_constant_type(self, context: HogQLContext) -> ConstantType:
-        database_field = self.resolve_database_field(context)
-        if isinstance(database_field, IntegerDatabaseField):
-            return IntegerType()
-        elif isinstance(database_field, FloatDatabaseField):
-            return FloatType()
-        elif isinstance(database_field, StringDatabaseField):
-            return StringType()
-        elif isinstance(database_field, BooleanDatabaseField):
-            return BooleanType()
-        elif isinstance(database_field, DateTimeDatabaseField):
-            return DateTimeType()
-        elif isinstance(database_field, DateDatabaseField):
-            return DateType()
-        return UnknownType()
+        table_type = self.table_type
+        while True:
+            if isinstance(table_type, SelectQueryAliasType):
+                table_type = table_type.select_query_type
+                continue
+
+            if isinstance(table_type, SelectQueryType):
+                field_type = table_type.columns.get(self.name)
+                if field_type is None:
+                    raise QueryError(f"Cant get column: {self.name}")
+
+                constant_type = field_type.resolve_constant_type(context)
+                if constant_type is None:
+                    raise QueryError(f"Constant type cant be resolved: {field_type}")
+
+                return constant_type
+
+            if isinstance(table_type, BaseTableType):
+                table: Table = table_type.resolve_database_table(context)
+                database_field = None
+                if table is not None:
+                    database_field = table.get_field(self.name)
+                if isinstance(database_field, DatabaseField):
+                    constant_type = database_field.get_constant_type()
+                    return constant_type
+                else:
+                    raise NotImplementedError(
+                        f"FieldType.resolve_constant_type, for BaseTableType: unknown database_field type: {str(database_field.__class__)}"
+                    )
+
+            if isinstance(table_type, SelectUnionQueryType):
+                table_type = table_type.types[0]
+                continue
+
+            if isinstance(table_type, SelectViewType):
+                table_type = table_type.select_query_type
+                continue
+
+            raise NotImplementedError(
+                f"FieldType.resolve_constant_type found unexpected table_type: {str(table_type.__class__)}"
+            )
```

Comment:
It feels like we shouldn't be doing this here... it should be more that we send it off to `table_or_select_or_alias_type.resolve_column_constant_type(self.name)`, and using class inheritance or whatnot, it'll recursively find its way to the deepest node.